### ISSUE TITLE
High Quality Image

### ISSUE THREAD

**Issue by sulaimanvesal** [Project-MONAI/GenerativeModels#374]
Hi all,

Thank you for this amazing efforts to put together LDM and DDM.

I have one question, on how to improve the quality of LDM. I tried to train LDM but the quality of generated images using LDM is pretty low. I saw also on the tutorials that for brain MRI the quality of images aren't good. 

Any insights on this would be great.

**Comment by xiaoiker** [comment]
Hi, in the tutorials the training samples are downsampled to low-quality ones. Besides, there are only trained for hundreds of epoches. So I think you just need to train the model with high-quality images,  make the model biger and train the model long eough.

**Comment by marksgraham** [comment]
Hi @sulaimanvesal 

You may also increase the quality of the latent model by using a perceptual loss and training adversarially. There is an example of how to train a VQ-GAN [here](https://github.com/Project-MONAI/GenerativeModels/tree/main/tutorials/generative/2d_vqgan).

**Comment by sulaimanvesal** [comment]
Thank you both for the suggestions. As @xiaoiker suggested, I tried with high image resolution and deeper models for both autoencoder and unet-diff (inspired by the config file from Model Zoo, for brain MRI synthesis. Both models trained for 1000 epochs each. 

@marksgraham The autoencoder trained with perceptual loss. 

I observed some improvements compared to initial attempts, but it's far behind the paper results. Perhaps, I need to increase the training samples further or maybe moving to 2D LDM model.

![image](https://user-images.githubusercontent.com/20819071/233315169-69f4440c-faa6-4b3f-a4cf-d9cecc0a5979.png)

**Comment by AbrahamGhavabesh** [comment]
@Warvito 
Hi, I try your model for training Autoencoderkl and Diffusion [model:](https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/2d_super_resolution/2d_stable_diffusion_v2_super_resolution.ipynb) for my personal program. It won't work for higher resolutions than 64*64(I test it on 256*256). Is there any help to guide me??? Thanks a lot

**Comment by sulaimanvesal** [comment]
Hi @Warvito,

Thank you for providing detailed information. I was able to obtain some decent prostate MRI images using AutoencoderKL, but as you suggested, I replaced it with 3D VQ_VAE. The image reconstruction quality of VQ_VAE is comparable to that of AutoencoderKL. I was hoping that the DiffusionUNet would generate higher quality images now, but despite the training curve indicating convergence, the output is just noise, as shown in the attached example.

![image](https://user-images.githubusercontent.com/20819071/235822491-71476079-2a22-41bc-aa31-df08b5143b61.png)


Regarding implementation, I followed the VQ-VAE using perceptual loss and adversarial training closely. I also looked into LDM-1 from Rambach et al. (appendix) for hyperparameter tuning. I thought that perhaps the UNet diffusion model needed to be changed, so I trained a few models with different configurations, but during inference, the models only produced noise. However, this behavior was not observed with AutoencoderKL. 

Any suggestion why VQ-VAE didn't work? do you have a tutorial for using 3D LDM with VQ-VAE rather than AutoencoderKL?

![image](https://user-images.githubusercontent.com/20819071/235821998-f1541015-bef8-41f2-8c59-6be6dbca79d6.png)

![image](https://user-images.githubusercontent.com/20819071/235822325-b7886e06-1ac1-429b-b0ee-6d176a67d334.png)

**Comment by marksgraham** [comment]
Hi @sulaimanvesal 

Would you be able to share your training code for the LDM training off the 3D VQ-VAE?

**Comment by sulaimanvesal** [comment]
Hi @marksgraham 

Here is the training code on my [onedrive](https://drive.google.com/file/d/11YFw6_OK2YRF2ojhnCkGzvdAFQC9LbTu/view?usp=sharing)

**Comment by sulaimanvesal** [comment]
Any update on the 3D LDM+VQ-VAE code review? I am still thinking how to improve this further.

**Comment by marksgraham** [comment]
> Any update on the 3D LDM+VQ-VAE code review? I am still thinking how to improve this further.

You need to grant me accesss to view the code - have requested.

**Comment by marksgraham** [comment]
Hi @sulaimanvesal 

I've taken a look at the code and can't see anything glaringly wrong that would produce those samples, but I have some thoughts. I have successfully trained a VQVAE LDM on 3d data with MONAI Generative, so know it is possible!

- Have you tried training a 3-layer VQVAE followed by a DDPM? In general, its easier to train a DDPM to sample correctly on a smaller latent space. The VQ-VAE recon quality won't be as good, but if you get some samples out that look like prostate MRIs it confirms there are no bugs in your overall pipeline, and getting it to work on a 2-layer VQ-VAE will just be a matter of tweaking things
- I would recommend trying to train/sample with the default epsilon prediction type, for now. In my experience v_prediction can make sampling worse. 
- You're UNet is quite a bit bigger than the example [here](https://github.com/Project-MONAI/GenerativeModels/blob/main/tutorials/generative/3d_ldm/3d_ldm_tutorial.py) - might be easier to train if you reduce the num_channels/head_channels?
